Improve binance spot trading logic

silva-fj opened the pull request:
Improving cross chain swaps by waiting until the deposit has been confirmed before attempt to create a market trade order on Binance. Additionaly, if the trade fails, the assets are transferred back to the OmniAccount

silva-fj committed:
feat(binance-api): add deposit history endpoint

silva-fj committed:
feat(cross-chain): verify deposits before trading

Add deposit confirmation logic to cross-chain executor to ensure funds
have been successfully received by Binance before attempting to create
market orders. Implements a polling mechanism with timeout to verify
deposit status, making the cross-chain swap process more robust and
reliable.

silva-fj committed:
updating status name

silva-fj committed:
Merge branch 'dev' into p-1466-improve-binance-spot-trading-logic

silva-fj committed:
feat(binance-api): add withdraw functionality

Implement the withdraw API endpoint for the Binance wallet API, allowing
the submission of withdrawal requests.

silva-fj committed:
Merge branch 'dev' into p-1466-improve-binance-spot-trading-logic

silva-fj committed:
feat(cross-chain): add withdrawal on failed trade

When a Binance spot trade order fails to be created, the system now
automatically withdraws the assets back to the Omni account instead of
just logging an error.

Kailai-Wang commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -647,6 +647,64 @@ impl<
+									continue;
+								},
+							};
+							if deposit.status == 1 && // 1 = success
+							   deposit.coin == binance_coin_name &&
+							   deposit.network == binance_network_info.network &&
+                               deposit_amount == amount_to_transfer
```
Hmm so if the user somehow makes two consectutive deposit with the same param (coin, network, amount) - can it be differentiated?

silva-fj replied:
There is a `txId` optional param, let me refactor the solana_client to return the signatures and pass it to the api. But I think we don't need to differentiate them, if the deposit was made the user cannot withdraw the assets. I've also included a filter for the source address to make sure the deposit is from the same user, I will push it soon

Kailai-Wang commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -669,6 +727,20 @@ impl<
+							.withdraw(
+								&binance_coin_name,
+								&from_address,
+								amount_to_transfer_decimal.to_string(),
```
Just to confirm, this is **with** multiplier, right?

"decimal" here just means the `Decimal` format, it doesn't mean it's e.g. `0.03` instead of `3 * (10^16)` (just example)

silva-fj replied:
Yes, this is with multiplier but looking at the API docs again, I think this should be the amount we get in pumpx_config, as binance expects a decimal number, so `0.03` in this example

Kailai-Wang commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -677,8 +749,6 @@ impl<
 						self.pumpx_api.cross_fail(&access_token, body).await.map_err(|_| {
 							log::error!("Failed to notify pumpx-signer");
 						})?;
-						// TODO: Figure out how to transfer back the asset to the omni account
```
There is another `TODO: Figure out how to transfer back the asset to the omni account` after the trade order, do we need to apply the withdraw too?

Could you please check that?

silva-fj replied:
done

Kailai-Wang reviewed (commented):
Thanks! I left a few comments

felixfaisal commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -647,6 +647,64 @@ impl<
+
+						if !deposit_confirmed {
+							debug!("Deposit not confirmed yet, waiting 5 seconds...");
+							sleep(Duration::from_secs(5)).await;
```
Is 5 seconds arbitrary? Finalisation delay could vary

silva-fj replied:
Yep

felixfaisal commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -647,6 +647,64 @@ impl<
+					}
+
+					if !deposit_confirmed {
+						log::error!("Deposit not confirmed within timeout period");
```
This would mean deposit not confirmed within timeout period even though we know deposit has been made on the chain and it's taking time for Binance to confirm because of re-orgs or something, should we have an exponential backoffs instead of arbitary 5s delay? wdyt?

silva-fj replied:
I was thinking about this, the problem is that according to the API docs there are a couple of cases where the deposit will not succeed, so we cannot just assume that It will be available eventually. In one case the deposit could be `rejected` (In this case I don't know what would happen, if binance just returns the assets to the source address or what), another case is `Wrong Deposit` (I don't know what that even means) and there is another `credited but cannot withdraw` which I don't know what we could do (this cannot be resolved with the API). In my experience deposits from Solana in binance are quite fast, but we'll need to revisit this once we start adding support for other networks.

The exponential backoffs sounds like a good improvement, we could get fancier in the next iterations 👍🏼 

I'm trying find something In the docs that would help to implement a more robust solution for these edge cases

Kailai-Wang replied:
Eventually it's a centralised service so yes it could be rejected (same for withdrawal - they could have arbitrary reasons to delay/reject it 😂)

These need time to improve, we don't expect a version that works for all corner cases. I'd say I'm fine if it can handle happy path + major error cases

felixfaisal replied:
Makes sense, but then i was wondering if deposit has been made on chain, and it timed out here because of the different cases you mentioned, then how will we be able to refund the user, because we probably don't have access to it anymore. 

I think we will get a better understanding once we start testing but something to look out for

felixfaisal commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -669,6 +727,20 @@ impl<
+								Some(&binance_network_info.network),
+							)
+							.await
+							.map_err(|_| {
```
I think it's important to segregate errors here, Like what are the possible errors? Could some types of errors be mitigated by performing retries (like network delay), or if the balance is too low(balance is locked and requires verification) then we have to readjust the amount, it'd be easier to debug and adjust if we segregate this,

silva-fj replied:
I agree, it would be awesome to know what the possible errors are but apart from regular network errors, there is no info in the docs about this

felixfaisal commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -647,6 +647,64 @@ impl<
+						};
+
+						// Check if there's a recent successful deposit
+						for deposit in deposit_history {
```
Does this also detects deposits that require more information from the account, if so can we programmatically fill the details of the deposit information that is required by KYC/AML of binance

silva-fj replied:
Can you elaborate a bit more?

felixfaisal replied:
When we receive a deposit on binance, we have to provide information on the depositor until then the deposit is not credited to your account, so does this detect deposits that are pending because of not providing depositor details. 

I'm not sure if you have tested it on production, if this didn't come up during your testing with prod binance api, then ig it's nothing to worry, but if you haven't tested then it's something to look out for.

silva-fj replied:
I see, I think from Binance perspective you (as the onwer of the binance account) are the depositor, so our account should have been checked already (?). But good point, we need to check all these scenarios when we start testing 👍🏼

silva-fj committed:
feat(binance-api): add timestamp filter to deposit history

silva-fj committed:
fix(cross-chain): validate deposit source address

silva-fj committed:
fix(cross-chain): use correct amount in withdrawal

silva-fj committed:
fix(cross-chain): improve error logging in withdrawal

silva-fj committed:
feat(cross-chain): track deposit transactions by ID

Add transaction ID tracking for cross-chain deposits to improve
reliability when checking deposit status. Modify Solana client to
return transaction signatures and pass them to Binance API to filter
deposit history by specific transactions. Also add handling for
rejected deposits with appropriate error reporting.

silva-fj committed:
Merge branch 'dev' into p-1466-improve-binance-spot-trading-logic

silva-fj committed:
feat(cross-chain): add asset recovery on failed trade

Implement withdrawal of assets back to the omni account when a
Binance trade fails.

Kailai-Wang reviewed (approved):
@silva-fj I'm happy with it mostly - if there're no major TODOs, I'm inclined to merging it and building a new image cc @BillyWooo

silva-fj commented:
> @silva-fj I'm happy with it mostly - if there're no major TODOs, I'm inclined to merging it and building a new image cc @BillyWooo

I agree, the sooner we start testing the better
